plugin/metrics.go

```diff
@@ -1,43 +1,201 @@
+	// 	err = errors.Join(err, exp.ForceFlush(context.Background()))
+	// }
+
+	return errors.Join(err, m.Shutdown(context.Background()))
+}
+
+func newMeterProvider(ctx context.Context) (shutDownMeterProvider, []metric.Exporter, error) {
+	exporters, err := newMetricsExporters(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if len(exporters) < 1 {
+		return &noopShutdownMeterProvider{noop.NewMeterProvider()}, nil, nil
+	}
+
+	options := []metric.Option{}
+	for _, exporter := range exporters {
+		options = append(options, metric.WithReader(metric.NewPeriodicReader(exporter)))
+	}
+
+	res, err := resource.Merge(
+		resource.Default(),
+		resource.NewSchemaless(
+			semconv.ServiceNameKey.String("kubo"),
+			semconv.ServiceNamespaceKey.String("ipfs"),
+			semconv.ServiceVersionKey.String("v0.20.0"),
+			semconv.TelemetrySDKLanguageGo,
+		),
+	)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	options = append(options, metric.WithResource(res))
+
+	return metric.NewMeterProvider(options...), exporters, nil
+}
+
+var (
+	errUnsupportedExporter      = errors.New("unknown or unsupported exporter")
+	errUnsupportedProtocol      = errors.New("unknown or unsupported OTLP protocol")
+	errBuildingOTLPgRPCExporter = errors.New("building OTLP gRPC exporter")
+	errBuildingOTLPHTTPExporter = errors.New("building OTLP HTTP exporter")
+	errBuildingStdoutExporter   = errors.New("building STDOUT exporter")
+)
+
+// newMetricExporters returns zero or more metrics exporters configured
+// strictly though environment variables.  Like kubo's OpenTelemetry
+// trace configuration.  See the README for set-up instructions.
+func newMetricsExporters(ctx context.Context) ([]metric.Exporter, error) {
+	for _, s := range os.Environ() {
+		if strings.HasPrefix(s, "OTEL") {
+			fmt.Println(s)
+		}
+	}
+
+	var exporters []metric.Exporter
+
+	exporterEnvVar := os.Getenv(envVarMetricsExporter)
+	if exporterEnvVar == "" {
+		return exporters, nil
+	}
+
+	for _, s := range strings.Split(exporterEnvVar, ",") {
+		if s != "otlp" {
+			return nil, fmt.Errorf("%w %s", errUnsupportedExporter, s)
+		}
+
+		protocol := otlpProtocolNameHTTP
+		if v := os.Getenv(envVarOTLPProtocol); v != "" {
+			protocol = v
+		}
+		if v := os.Getenv(envVarOTLPMetricsProtocol); v != "" {
+			protocol = v
+		}
+
+		switch protocol {
+		case otlpProtocolNameHTTP:
+			exporter, err := otlpmetrichttp.New(ctx)
+			if err != nil {
+				return nil, errors.Join(errBuildingOTLPHTTPExporter, err)
+			}
+			exporters = append(exporters, exporter)
+		case otlpProtocolNameGRPC:
+			exporter, err := otlpmetricgrpc.New(ctx)
+			if err != nil {
+				return nil, errors.Join(errBuildingOTLPgRPCExporter, err)
+			}
+			exporters = append(exporters, exporter)
+		default:
+			return nil, fmt.Errorf("%w %s", errUnsupportedProtocol, protocol)
+		}
+	}
+
+	stdoutExporter, err := stdoutmetric.New()
+	if err != nil {
+		return nil, errors.Join(errBuildingStdoutExporter)
+	}
+
+	_ = stdoutExporter
```

Comment:
add that construction as part of the commented out code? no need to instantiate that if we won't use it